zebra/zebra_evpn_neigh.c

```diff
@@ -0,0 +1,2452 @@
+	if (!vlan_if)
+		return -1;
+
+	flags = DPLANE_NTF_EXT_LEARNED;
+	if (n->flags & ZEBRA_NEIGH_ROUTER_FLAG)
+		flags |= DPLANE_NTF_ROUTER;
+	ZEBRA_NEIGH_SET_ACTIVE(n);
+
+	dplane_rem_neigh_add(vlan_if, &n->ip, &n->emac, flags, was_static);
+
+	return ret;
+}
+
+/*
+ * Install neighbor hash entry - called upon access VLAN change.
+ */
+void zebra_evpn_install_neigh_hash(struct hash_bucket *bucket, void *ctxt)
+{
+	zebra_neigh_t *n;
+	struct neigh_walk_ctx *wctx = ctxt;
+
+	n = (zebra_neigh_t *)bucket->data;
+
+	if (CHECK_FLAG(n->flags, ZEBRA_NEIGH_REMOTE))
+		zebra_evpn_rem_neigh_install(wctx->zevpn, n,
+					     false /*was_static*/);
+}
+
+/*
+ * Callback to allocate neighbor hash entry.
+ */
+static void *zebra_evpn_neigh_alloc(void *p)
+{
+	const zebra_neigh_t *tmp_n = p;
+	zebra_neigh_t *n;
+
+	n = XCALLOC(MTYPE_NEIGH, sizeof(zebra_neigh_t));
+	*n = *tmp_n;
+
+	return ((void *)n);
+}
+
+static void zebra_evpn_local_neigh_ref_mac(zebra_neigh_t *n,
+					   struct ethaddr *macaddr,
+					   zebra_mac_t *mac,
+					   bool send_mac_update)
+{
+	char macbuf[ETHER_ADDR_STRLEN];
+	char ipbuf[INET6_ADDRSTRLEN];
+	bool old_static;
+	bool new_static;
+
+	memcpy(&n->emac, macaddr, ETH_ALEN);
+	n->mac = mac;
+
+	/* Link to new MAC */
+	if (!mac)
+		return;
+
+	listnode_add_sort(mac->neigh_list, n);
+	if (n->flags & ZEBRA_NEIGH_ALL_PEER_FLAGS) {
+		old_static = zebra_evpn_mac_is_static(mac);
+		++mac->sync_neigh_cnt;
+		new_static = zebra_evpn_mac_is_static(mac);
+		if (IS_ZEBRA_DEBUG_EVPN_MH_NEIGH)
+			zlog_debug(
+				"sync-neigh ref mac vni %u ip %s mac %s ref %d",
+				n->zevpn->vni,
+				ipaddr2str(&n->ip, ipbuf, sizeof(ipbuf)),
+				prefix_mac2str(&n->emac, macbuf,
+					       sizeof(macbuf)),
+				mac->sync_neigh_cnt);
+		if ((old_static != new_static) && send_mac_update)
+			/* program the local mac in the kernel */
+			zebra_evpn_sync_mac_dp_install(
+				mac, false /*set_inactive*/,
+				false /*force_clear_static*/, __func__);
+	}
+}
+
+/* sync-path that is active on an ES peer */
+static void zebra_evpn_sync_neigh_dp_install(zebra_neigh_t *n,
+					     bool set_inactive,
+					     bool force_clear_static,
+					     const char *caller)
+{
+	char macbuf[ETHER_ADDR_STRLEN];
+	char ipbuf[INET6_ADDRSTRLEN];
+	struct zebra_ns *zns;
+	struct interface *ifp;
+	bool set_static;
+	bool set_router;
+
+	zns = zebra_ns_lookup(NS_DEFAULT);
+	ifp = if_lookup_by_index_per_ns(zns, n->ifindex);
+	if (!ifp) {
+		if (IS_ZEBRA_DEBUG_EVPN_MH_NEIGH)
+			zlog_debug(
+				"%s: dp-install sync-neigh vni %u ip %s mac %s if %d f 0x%x skipped",
+				caller, n->zevpn->vni,
+				ipaddr2str(&n->ip, ipbuf, sizeof(ipbuf)),
+				prefix_mac2str(&n->emac, macbuf,
+					       sizeof(macbuf)),
+				n->ifindex, n->flags);
+		return;
+	}
+
+	if (force_clear_static)
+		set_static = false;
+	else
+		set_static = zebra_evpn_neigh_is_static(n);
+
+	set_router = !!CHECK_FLAG(n->flags, ZEBRA_NEIGH_ROUTER_FLAG);
+
+	/* XXX - this will change post integration with the new kernel */
```

Comment:
Maybe we should add a deprecation warning? Or check the kernel version somehow?

I fear this might break in the future without notice... or am I making it bigger than it is?